File: src/geoseal_v2.py
"""
GeoSeal v2: Mixed-Curvature Geometric Access Control Kernel
============================================================

Product manifold H^a x S^b x R^c where:
- Hyperbolic (H^a): hierarchy, trust zones, boundary quarantine
- Spherical  (S^b): tongue phase discipline, cyclic role coherence
- Gaussian   (R^c): retrieval uncertainty, memory write gating

Each agent carries three coordinate families:
  u in B^n   (Poincare ball position)     -- hierarchy / containment
  p in S^1   (phase as [cos t, sin t])    -- tongue discipline
  (mu, s^2)  (diagonal Gaussian)          -- retrieval confidence

Scoring fuses three independent geometry scores:
  trust = w_H * s_H + w_S * s_S + w_G * s_G
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Dict, List, Literal, Optional, Tuple

from src.geoseal import (
    TONGUE_PHASES,
    SUSPICION_DECAY,
    SUSPICION_THRESHOLD,
    QUARANTINE_CONSENSUS,
    TRUST_DENOMINATOR,
    hyperbolic_distance,
    phase_deviation,
    clamp_to_ball,
)

# ---------------------------------------------------------------------------
# Fusion weights
# ---------------------------------------------------------------------------

DEFAULT_WEIGHTS: Tuple[float, float, float] = (0.4, 0.35, 0.25)  # wH, wS, wG
QUARANTINE_TRUST_THRESHOLD = 0.3
MEMORY_WRITE_THRESHOLD = 0.7

Action = Literal["ALLOW", "QUARANTINE", "DENY"]


# ---------------------------------------------------------------------------
# Mixed-geometry Agent (v2)
# ---------------------------------------------------------------------------


@dataclass
class MixedAgent:
    """Agent with hyperbolic position, spherical phase, and Gaussian uncertainty."""

    id: str
    position: List[float]
    phase: Optional[float]
    sigma: float = 0.0
    tongue: Optional[str] = None
    suspicion_count: Dict[str, float] = field(default_factory=dict)
    is_quarantined: bool = False
    trust_score: float = 0.5
    score_hyperbolic: float = 0.0
    score_phase: float = 0.0
    score_certainty: float = 0.0

    @property
    def phase_vec(self) -> Tuple[float, float]:
        """Phase as [cos theta, sin theta]."""
        if self.phase is not None:
            return (math.cos(self.phase), math.sin(self.phase))
        return (0.0, 0.0)


# ---------------------------------------------------------------------------
# Individual geometry scores
# ---------------------------------------------------------------------------


def score_hyperbolic(a: MixedAgent, b: MixedAgent) -> float:
    """s_H = 1 / (1 + d_H). High when close in Poincare ball."""
    d_h = hyperbolic_distance(a.position, b.position)
    return 1.0 / (1.0 + d_h)


def score_phase(a: MixedAgent, b: MixedAgent) -> float:
    """s_S = 1 - phaseDeviation. High when same tongue phase."""
    return 1.0 - phase_deviation(a.phase, b.phase)


def score_certainty(b: MixedAgent) -> float:
    """s_G = 1 / (1 + sigma). High when low uncertainty."""
    return 1.0 / (1.0 + b.sigma)


# ---------------------------------------------------------------------------
# Product manifold fusion
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class FusedScore:
    trust: float
    s_h: float
    s_s: float
    s_g: float
    anomaly: bool
    action: Action


def fuse_scores(
    anchor: MixedAgent,
    candidate: MixedAgent,
    weights: Tuple[float, float, float] = DEFAULT_WEIGHTS,
) -> FusedScore:
    """Fuse three geometry scores into a single trust value."""
    w_h, w_s, w_g = weights
    s_h = score_hyperbolic(anchor, candidate)
    s_s = score_phase(anchor, candidate)
    s_g = score_certainty(candidate)

    trust = w_h * s_h + w_s * s_s + w_g * s_g
    anomaly = s_s < 0.5 or s_g < 0.5

    if trust >= MEMORY_WRITE_THRESHOLD:
        action: Action = "ALLOW"
    elif trust >= QUARANTINE_TRUST_THRESHOLD:
        action = "QUARANTINE"
    else:
        action = "DENY"

    return FusedScore(trust=trust, s_h=s_h, s_s=s_s, s_g=s_g, anomaly=anomaly, action=action)


# ---------------------------------------------------------------------------
# Suspicion (same algorithm as v1)
# ---------------------------------------------------------------------------


def update_suspicion_v2(agent: MixedAgent, neighbor_id: str, is_anomaly: bool) -> None:
    if is_anomaly:
        agent.suspicion_count[neighbor_id] = (
            agent.suspicion_count.get(neighbor_id, 0) + 1
        )
    else:
        agent.suspicion_count[neighbor_id] = max(
            0, agent.suspicion_count.get(neighbor_id, 0) - SUSPICION_DECAY
        )

    suspicious_neighbors = sum(
        1 for c in agent.suspicion_count.values() if c >= SUSPICION_THRESHOLD
    )
    agent.is_quarantined = suspicious_neighbors >= QUARANTINE_CONSENSUS

    total_suspicion = sum(agent.suspicion_count.values())
    agent.trust_score = max(0, 1.0 - total_suspicion / TRUST_DENOMINATOR)


# ---------------------------------------------------------------------------
# v2 Repulsion
# ---------------------------------------------------------------------------


def compute_repel_force_v2(
    agent_a: MixedAgent,
    agent_b: MixedAgent,
    anchor: Optional[MixedAgent] = None,
    base_strength: float = 1.0,
    weights: Tuple[float, float, float] = DEFAULT_WEIGHTS,
) -> Tuple[List[float], float, bool, FusedScore]:
    """v2 repulsion with uncertainty amplification.

    Returns (force, amplification, anomaly_flag, fused_score).
    """
    d_h = hyperbolic_distance(agent_a.position, agent_b.position)
    base_repulsion = base_strength / (d_h + 1e-6)

    amplification = 1.0
    anomaly_flag = False

    if agent_b.phase is None:
        amplification = 2.0
        anomaly_flag = True
    elif agent_a.phase is not None:
        deviation = phase_deviation(agent_a.phase, agent_b.phase)
        if d_h < 1.0 and deviation > 0.5:
            amplification = 1.5 + deviation
            anomaly_flag = True

    if agent_b.is_quarantined:
        amplification *= 1.5

    # v2: uncertainty amplification
    if agent_b.sigma > 0.5:
        amplification += 0.5
        anomaly_flag = True

    # v2: fused score
    # Only apply fused anomaly when source has valid phase; a null-phase source
    # would always produce sS=0, falsely flagging legitimate targets.
    ref = anchor or agent_a
    fused = fuse_scores(ref, agent_b, weights)
    if fused.anomaly and ref.phase is not None:
        amplification += 0.25
        anomaly_flag = True

    direction = [a - b for a, b in zip(agent_a.position, agent_b.position)]
    force = [d * base_repulsion * amplification for d in direction]

    agent_b.score_hyperbolic = fused.s_h
    agent_b.score_phase = fused.s_s
    agent_b.score_certainty = fused.s_g

    return force, amplification, anomaly_flag, fused


# ---------------------------------------------------------------------------
# v2 Swarm dynamics
# ---------------------------------------------------------------------------


def swarm_step_v2(
    agents: List[MixedAgent],
    drift_rate: float = 0.01,
    sigma_decay: float = 0.01,
    weights: Tuple[float, float, float] = DEFAULT_WEIGHTS,
) -> List[MixedAgent]:
    """One v2 swarm update step with uncertainty evolution."""
    n = len(agents)
    if n == 0:
        return agents

    dim = len(agents[0].position)

    for i in range(n):
        net_force = [0.0] * dim

        for j in range(n):
            if i == j:
                continue

            force, _amp, anomaly, _fused = compute_repel_force_v2(
                agents[i], agents[j], None, 1.0, weights
            )
            for k in range(dim):
                net_force[k] += force[k]

            update_suspicion_v2(agents[j], agents[i].id, anomaly)

        # Apply force
        for k in range(dim):
            agents[i].position[k] += net_force[k] * drift_rate
        agents[i].position = clamp_to_ball(agents[i].position, 0.99)

        # v2: uncertainty evolution
        # Sigma is driven by how much others flag THIS agent (suspicion_count),
        # not by how many anomalies it detects on others.
        total_incoming = sum(agents[i].suspicion_count.values())
        if total_incoming > 3:
            agents[i].sigma = min(10.0, agents[i].sigma + sigma_decay * 2)
        else:
            agents[i].sigma = max(0, agents[i].sigma - sigma_decay)

    return agents


def run_swarm_v2(
    agents: List[MixedAgent],
    num_steps: int = 10,
    drift_rate: float = 0.01,
    sigma_decay: float = 0.01,
    weights: Tuple[float, float, float] = DEFAULT_WEIGHTS,
) -> List[MixedAgent]:
    """Run multiple v2 swarm steps."""
    for _ in range(num_steps):
        swarm_step_v2(agents, drift_rate, sigma_decay, weights)
    return agents


# ---------------------------------------------------------------------------
# Batch scoring
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class ScoredCandidate:
    id: str
    trust: float
    action: Action
    s_h: float
    s_s: float
    s_g: float
    is_quarantined: bool
    sigma: float


def score_all_candidates(
    anchors: List[MixedAgent],
    candidates: List[MixedAgent],
    weights: Tuple[float, float, float] = DEFAULT_WEIGHTS,
) -> List[ScoredCandidate]:
    """Score all candidates against tongue anchors. Returns sorted by trust desc."""
    results: List[ScoredCandidate] = []

    for candidate in candidates:
        best: Optional[FusedScore] = None
        for anchor in anchors:
            fused = fuse_scores(anchor, candidate, weights)
            if best is None or fused.trust > best.trust:
                best = fused

        if best is not None:
            results.append(
                ScoredCandidate(
                    id=candidate.id,
                    trust=best.trust,
                    action=best.action,
                    s_h=best.s_h,
                    s_s=best.s_s,
                    s_g=best.s_g,
                    is_quarantined=candidate.is_quarantined,
                    sigma=candidate.sigma,
                )
            )

    results.sort(key=lambda x: x.trust, reverse=True)
    return results

File: src/geoseal-v2.ts
/**
 * @file geoseal-v2.ts
 * @module geoseal-v2
 * @layer Layer 9, Layer 12, Layer 13
 * @version 2.0.0
 *
 * GeoSeal v2: Mixed-Curvature Geometric Access Control Kernel
 *
 * Extends GeoSeal v1 with a product manifold H^a x S^b x R^c where:
 * - Hyperbolic (H^a): hierarchy, trust zones, boundary quarantine
 * - Spherical  (S^b): tongue phase discipline, cyclic role coherence
 * - Gaussian   (R^c): retrieval uncertainty, memory write gating
 *
 * Each agent carries three coordinate families:
 *   u ∈ B^n   (Poincaré ball position)     — hierarchy / containment
 *   p ∈ S^1   (phase as [cos θ, sin θ])    — tongue discipline
 *   (μ, σ²)   (diagonal Gaussian)          — retrieval confidence
 *
 * Scoring fuses three independent geometry scores:
 *   trust = w_H · s_H + w_S · s_S + w_G · s_G
 *
 * where:
 *   s_H = 1 / (1 + d_H)               — hyperbolic proximity
 *   s_S = 1 - phaseDeviation           — phase consistency
 *   s_G = 1 / (1 + trace(Σ))          — uncertainty (low = trustworthy)
 *
 * Quarantine triggers when fused trust drops below threshold OR when
 * spatial consensus (3+ neighbors) flags the agent.
 */

import {
  hyperbolicDistance,
  phaseDeviation,
  clampToBall,
} from './harmonic/hyperbolic.js';
import {
  TONGUE_PHASES,
  SUSPICION_DECAY,
  SUSPICION_THRESHOLD,
  QUARANTINE_CONSENSUS,
  TRUST_DENOMINATOR,
} from './geoseal.js';

// Re-export v1 constants for convenience
export { TONGUE_PHASES, SUSPICION_THRESHOLD, QUARANTINE_CONSENSUS };

// ═══════════════════════════════════════════════════════════════
// Fusion weights (tunable hyperparameters)
// ═══════════════════════════════════════════════════════════════

export interface FusionWeights {
  /** Weight for hyperbolic proximity score */
  wH: number;
  /** Weight for spherical phase consistency score */
  wS: number;
  /** Weight for Gaussian certainty score */
  wG: number;
}

export const DEFAULT_FUSION_WEIGHTS: FusionWeights = {
  wH: 0.4,
  wS: 0.35,
  wG: 0.25,
};

/** Quarantine threshold for fused trust score */
export const QUARANTINE_TRUST_THRESHOLD = 0.3;

/** Memory write threshold (only chunks above this get promoted) */
export const MEMORY_WRITE_THRESHOLD = 0.7;

// ═══════════════════════════════════════════════════════════════
// Mixed-geometry Agent (v2)
// ═══════════════════════════════════════════════════════════════

export interface MixedAgent {
  id: string;

  // --- Hyperbolic component (H^a) ---
  /** Embedding position in Poincaré ball */
  position: number[];

  // --- Spherical component (S^1) ---
  /** Tongue phase angle (null if unknown/rogue) */
  phase: number | null;
  /** Phase as [cos θ, sin θ] for computation (auto-derived) */
  phaseVec: [number, number];

  // --- Gaussian component ---
  /** Uncertainty variance (diagonal; scalar per dimension) */
  sigma: number;

  // --- Metadata ---
  tongue?: string;
  suspicion_count: Map<string, number>;
  is_quarantined: boolean;
  trust_score: number;

  // --- v2 scoring breakdown ---
  score_hyperbolic: number;
  score_phase: number;
  score_certainty: number;
}

/**
 * Create a v2 mixed-geometry agent.
 *
 * @param id - Agent identifier
 * @param position - Poincaré ball embedding
 * @param phase - Tongue phase (null = rogue)
 * @param sigma - Uncertainty variance (0 = perfectly certain)
 * @param tongue - Optional tongue code
 * @param trust - Initial trust score
 */
export function createMixedAgent(
  id: string,
  position: number[],
  phase: number | null,
  sigma: number = 0.0,
  tongue?: string,
  trust: number = 0.5
): MixedAgent {
  return {
    id,
    position: [...position],
    phase,
    phaseVec: phase !== null ? [Math.cos(phase), Math.sin(phase)] : [0, 0],
    sigma,
    tongue,
    suspicion_count: new Map(),
    is_quarantined: false,
    trust_score: trust,
    score_hyperbolic: 0,
    score_phase: 0,
    score_certainty: 0,
  };
}

// ═══════════════════════════════════════════════════════════════
// Individual geometry scores
// ═══════════════════════════════════════════════════════════════

/**
 * Hyperbolic proximity score: s_H = 1 / (1 + d_H)
 * High when agents are close in the Poincaré ball.
 */
export function scoreHyperbolic(a: MixedAgent, b: MixedAgent): number {
  const dH = hyperbolicDistance(a.position, b.position);
  return 1.0 / (1.0 + dH);
}

/**
 * Spherical phase consistency score: s_S = 1 - phaseDeviation
 * High when agents share the same tongue phase.
 */
export function scorePhase(a: MixedAgent, b: MixedAgent): number {
  return 1.0 - phaseDeviation(a.phase, b.phase);
}

/**
 * Gaussian certainty score: s_G = 1 / (1 + σ²)
 * High when the agent has low uncertainty.
 * Scored for agent b (the candidate being evaluated).
 */
export function scoreCertainty(b: MixedAgent): number {
  return 1.0 / (1.0 + b.sigma);
}

// ═══════════════════════════════════════════════════════════════
// Product manifold fusion
// ═══════════════════════════════════════════════════════════════

export interface FusedScore {
  /** Combined trust score in [0, 1] */
  trust: number;
  /** Individual geometry scores */
  sH: number;
  sS: number;
  sG: number;
  /** Anomaly flag (any geometry triggered) */
  anomaly: boolean;
  /** Recommended action */
  action: 'ALLOW' | 'QUARANTINE' | 'DENY';
}

/**
 * Fuse three geometry scores into a single trust value.
 *
 * trust = w_H * s_H + w_S * s_S + w_G * s_G
 *
 * Actions:
 * - ALLOW:      trust >= MEMORY_WRITE_THRESHOLD
 * - QUARANTINE: QUARANTINE_TRUST_THRESHOLD <= trust < MEMORY_WRITE_THRESHOLD
 * - DENY:       trust < QUARANTINE_TRUST_THRESHOLD
 *
 * @param anchor - Reference agent (e.g., tongue agent)
 * @param candidate - Agent being evaluated
 * @param weights - Fusion weights (default: 0.4/0.35/0.25)
 */
export function fuseScores(
  anchor: MixedAgent,
  candidate: MixedAgent,
  weights: FusionWeights = DEFAULT_FUSION_WEIGHTS
): FusedScore {
  const sH = scoreHyperbolic(anchor, candidate);
  const sS = scorePhase(anchor, candidate);
  const sG = scoreCertainty(candidate);

  const trust = weights.wH * sH + weights.wS * sS + weights.wG * sG;

  // Anomaly: low phase consistency or high uncertainty
  const anomaly = sS < 0.5 || sG < 0.5;

  let action: 'ALLOW' | 'QUARANTINE' | 'DENY';
  if (trust >= MEMORY_WRITE_THRESHOLD) {
    action = 'ALLOW';
  } else if (trust >= QUARANTINE_TRUST_THRESHOLD) {
    action = 'QUARANTINE';
  } else {
    action = 'DENY';
  }

  return { trust, sH, sS, sG, anomaly, action };
}

// ═══════════════════════════════════════════════════════════════
// v2 Suspicion (same algorithm as v1, applied to MixedAgent)
// ═══════════════════════════════════════════════════════════════

export function updateSuspicionV2(
  agent: MixedAgent,
  neighbor_id: string,
  is_anomaly: boolean
): void {
  if (is_anomaly) {
    const current = agent.suspicion_count.get(neighbor_id) || 0;
    agent.suspicion_count.set(neighbor_id, current + 1);
  } else {
    const current = agent.suspicion_count.get(neighbor_id) || 0;
    agent.suspicion_count.set(neighbor_id, Math.max(0, current - SUSPICION_DECAY));
  }

  let suspicious_neighbors = 0;
  for (const count of agent.suspicion_count.values()) {
    if (count >= SUSPICION_THRESHOLD) suspicious_neighbors++;
  }
  agent.is_quarantined = suspicious_neighbors >= QUARANTINE_CONSENSUS;

  let total_suspicion = 0;
  for (const count of agent.suspicion_count.values()) {
    total_suspicion += count;
  }
  agent.trust_score = Math.max(0, 1.0 - total_suspicion / TRUST_DENOMINATOR);
}

// ═══════════════════════════════════════════════════════════════
// v2 Repulsion (enhanced with uncertainty amplification)
// ═══════════════════════════════════════════════════════════════

export interface RepulsionResultV2 {
  force: number[];
  amplification: number;
  anomaly_flag: boolean;
  fused: FusedScore;
}

/**
 * v2 repulsion force: includes uncertainty in amplification.
 *
 * Amplification rules (additive):
 * - v1 phase rules still apply (null=2.0x, mismatch=1.5x+dev, quarantine=1.5x)
 * - High uncertainty (σ > 0.5) adds +0.5x amplification
 * - Fused anomaly adds +0.25x
 */
export function computeRepelForceV2(
  agent_a: MixedAgent,
  agent_b: MixedAgent,
  anchor: MixedAgent | null = null,
  base_strength: number = 1.0,
  weights: FusionWeights = DEFAULT_FUSION_WEIGHTS
): RepulsionResultV2 {
  const d_H = hyperbolicDistance(agent_a.position, agent_b.position);
  const base_repulsion = base_strength / (d_H + 1e-6);

  // v1 phase-based amplification
  let amplification = 1.0;
  let anomaly_flag = false;

  if (agent_b.phase === null) {
    amplification = 2.0;
    anomaly_flag = true;
  } else if (agent_a.phase !== null) {
    const deviation = phaseDeviation(agent_a.phase, agent_b.phase);
    if (d_H < 1.0 && deviation > 0.5) {
      amplification = 1.5 + deviation;
      anomaly_flag = true;
    }
  }

  if (agent_b.is_quarantined) {
    amplification *= 1.5;
  }

  // v2: uncertainty amplification
  if (agent_b.sigma > 0.5) {
    amplification += 0.5;
    anomaly_flag = true;
  }

  // v2: fused score (use anchor if provided, else agent_a)
  // Only apply fused anomaly amplification when the source has a valid phase;
  // a null-phase source would always produce sS=0, falsely flagging legit targets.
  const ref = anchor || agent_a;
  const fused = fuseScores(ref, agent_b, weights);

  if (fused.anomaly && ref.phase !== null) {
    amplification += 0.25;
    anomaly_flag = true;
  }

  // Force vector
  const dim = agent_a.position.length;
  const force = new Array<number>(dim);
  for (let i = 0; i < dim; i++) {
    const direction = agent_a.position[i] - agent_b.position[i];
    force[i] = direction * base_repulsion * amplification;
  }

  // Store score breakdown on the candidate
  agent_b.score_hyperbolic = fused.sH;
  agent_b.score_phase = fused.sS;
  agent_b.score_certainty = fused.sG;

  return { force, amplification, anomaly_flag, fused };
}

// ═══════════════════════════════════════════════════════════════
// v2 Swarm dynamics
// ═══════════════════════════════════════════════════════════════

/**
 * Run one v2 swarm update step.
 *
 * Same structure as v1 but uses mixed-geometry repulsion and
 * optionally applies uncertainty decay (σ decreases when an agent
 * consistently matches its neighbors, increases under anomaly).
 */
export function swarmStepV2(
  agents: MixedAgent[],
  drift_rate: number = 0.01,
  sigma_decay: number = 0.01,
  weights: FusionWeights = DEFAULT_FUSION_WEIGHTS
): MixedAgent[] {
  const n = agents.length;
  const dim = agents[0]?.position.length ?? 0;

  for (let i = 0; i < n; i++) {
    const net_force = new Array<number>(dim).fill(0);

    for (let j = 0; j < n; j++) {
      if (i === j) continue;

      const result = computeRepelForceV2(agents[i], agents[j], null, 1.0, weights);

      for (let k = 0; k < dim; k++) {
        net_force[k] += result.force[k];
      }

      // Update suspicion on the TARGET (j) from the SOURCE (i)
      updateSuspicionV2(agents[j], agents[i].id, result.anomaly_flag);
    }

    // Apply force with drift
    for (let k = 0; k < dim; k++) {
      agents[i].position[k] += net_force[k] * drift_rate;
    }
    agents[i].position = clampToBall(agents[i].position, 0.99);

    // v2: uncertainty evolution
    // Sigma is driven by how much others flag THIS agent (suspicion_count),
    // not by how many anomalies it detects on others.
    let total_incoming_suspicion = 0;
    for (const count of agents[i].suspicion_count.values()) {
      total_incoming_suspicion += count;
    }
    if (total_incoming_suspicion > 3) {
      // Others are flagging this agent: uncertainty grows
      agents[i].sigma = Math.min(10.0, agents[i].sigma + sigma_decay * 2);
    } else {
      // Not being flagged: uncertainty decays
      agents[i].sigma = Math.max(0, agents[i].sigma - sigma_decay);
    }
  }

  return agents;
}

/**
 * Run multiple v2 swarm steps.
 */
export function runSwarmV2(
  agents: MixedAgent[],
  num_steps: number = 10,
  drift_rate: number = 0.01,
  sigma_decay: number = 0.01,
  weights: FusionWeights = DEFAULT_FUSION_WEIGHTS
): MixedAgent[] {
  for (let step = 0; step < num_steps; step++) {
    swarmStepV2(agents, drift_rate, sigma_decay, weights);
  }
  return agents;
}

// ═══════════════════════════════════════════════════════════════
// v2 Batch scoring (for RAG result reranking)
// ═══════════════════════════════════════════════════════════════

export interface ScoredCandidate {
  id: string;
  trust: number;
  action: 'ALLOW' | 'QUARANTINE' | 'DENY';
  sH: number;
  sS: number;
  sG: number;
  is_quarantined: boolean;
  sigma: number;
}

/**
 * Score all candidates against a set of tongue anchors.
 *
 * For each candidate, computes the best fused score across all anchors
 * (i.e., the tongue that trusts this candidate the most).
 *
 * @param anchors - Tongue agents (trusted references)
 * @param candidates - Retrieval/memory agents to score
 * @param weights - Fusion weights
 * @returns Scored candidates sorted by trust (descending)
 */
export function scoreAllCandidates(
  anchors: MixedAgent[],
  candidates: MixedAgent[],
  weights: FusionWeights = DEFAULT_FUSION_WEIGHTS
): ScoredCandidate[] {
  const results: ScoredCandidate[] = [];

  for (const candidate of candidates) {
    // Best score across all anchors
    let bestFused: FusedScore | null = null;

    for (const anchor of anchors) {
      const fused = fuseScores(anchor, candidate, weights);
      if (!bestFused || fused.trust > bestFused.trust) {
        bestFused = fused;
      }
    }

    if (bestFused) {
      results.push({
        id: candidate.id,
        trust: bestFused.trust,
        action: bestFused.action,
        sH: bestFused.sH,
        sS: bestFused.sS,
        sG: bestFused.sG,
        is_quarantined: candidate.is_quarantined,
        sigma: candidate.sigma,
      });
    }
  }

  // Sort by trust descending
  results.sort((a, b) => b.trust - a.trust);
  return results;
}
